### ISSUE TITLE
VM may throw when stored value exceeds 32 bytes 

### ISSUE THREAD

**Issue by davidmurdoch** [ethereumjs/ethereumjs-vm#464]
The VM currently allows values larger than 32 bytes to be stored.

I think the cause is because `SSTORE` does not limit the size of the value, as seen here:
https://github.com/ethereumjs/ethereumjs-vm/blob/0162d471957a1a31663b64eac7f14bff5a8e29f6/lib/vm/opFns.js#L540

However, the `MSTORE` opcode _does_ limit the size of the value that can be stored:
https://github.com/ethereumjs/ethereumjs-vm/blob/0162d471957a1a31663b64eac7f14bff5a8e29f6/lib/vm/opFns.js#L497

An issue was filed over in Ganache: https://github.com/trufflesuite/ganache-cli/issues/626 where a user was testing this edge case (though they were expecting to get a value back that was greater than a max `uint256`).

I'm not sure what the proper way to handle this is.

To test:

 * Store the following contract:

```
pragma solidity 0.4.25;

contract MyContract {
    uint public balance;
    function deposit() external payable {
        balance = address(this).balance;
    }
}
```
 * seed your account with:

`account.balance = new BN(2).pow(new BN(256)).addn(2);`

 * send the following transaction _twice_:

`{to: contractAddress, from: account.address, value: new BN(2).pow(new BN(255))}`

Then read the contract's public field `balance` (this should call `MSTORE`). This should trigger the exception. Interesting note is that just getting the balance of the contract via its address (like via `eth_getBalance` over JSON-RPC) works and is accurate (it will be equal to `new BN(2).pow(new BN(256)).addn(2)`).

Let me know if you need any more info!

**Comment by davidmurdoch** [comment]
I have a test case to demonstrate the issue here:

Failing test case: https://github.com/trufflesuite/ganache-core/blob/a62740aff7a8910e2936e1d891fd799cd6fc6477/test/contract-storage.js#L53-L54

Branch: https://github.com/trufflesuite/ganache-core/tree/contract-storage-overflow

Easiest way to reproduce is to clone the repo and switch to the `contract-storage-overflow` branch then run:

`npm install`
`npm run _mocha`

**Comment by s1na** [comment]
The test case calls `stateManager.putAccount` directly and sets a value wider than 32 bytes in the storage. Just to be sure, this is not happening during *normal* operations of the VM, right?

I'm not sure what we can do on our end. I could add this before setting the value in `SSTORE`:

```js
if (val.byteLength() > 32) trap(ERROR.OUT_OF_RANGE)
```

Do you think that could help? or do you have other suggestions?

**Comment by davidmurdoch** [comment]
In case you missed it, [this issue](https://github.com/trufflesuite/ganache-cli/issues/626) outlines a real-world case where someone managed to crash the VM via ganache by hitting the 32 byte conversion issue. The test is just a quick and dirty way to get the account's state in a place where we can trigger the crash without having to have an understanding of ganache-core internals and options.

---

I'm not sure what the proper behavior is here. Reading through geth source code it looks like the EVM should silently truncate the data:

`SSTORE`:
https://github.com/ethereum/go-ethereum/blob/391d4cb9b53eb9f71862cbb8a6cca332d5a76f94/core/vm/instructions.go#L636-L643

calls `BigToHash`:
https://github.com/ethereum/go-ethereum/blob/1a3e25e4c11d09aabd77402585412f1577251f49/common/types.go#L33-L59

Which uses their `Hash` type which is a fixed-length 32-byte array. The comments in their code indicate that `If [the byte array] is larger than [length of the hash], [the byte array] will be cropped from the left.`

If we were to use Geth as a reference implementation, then instead of throwing an `OUT_OF_RANGE` error maybe something like the following would be appropriate:

```
// Untested
var value;
if (val.isZero()){
  value = Buffer.from([])
} else if (val.bitLength() > 256 ) {
  // if the length is greater than 256 bits we need to truncate
  value = val.toArrayLike(Buffer, 'be').slice(0, 32)
} else {
  // TODO: Should this be padded to 32 bytes? I think not?
  value = val.toArrayLike(Buffer, 'be')
}
```

This snippet could be used anywhere `toArrayLike` is used in the context of setting storage for both keys and values.

**Comment by s1na** [comment]
To clarify, I didn't mean to discount the crash. What I meant was that, since the evm works on 32 byte words, this error should not occur if a larger value is not injected directly to the storage. But I'm fully in agreement that when a larger than max value is injected, the vm should not crash.

Py-evm also seems to [throw](https://github.com/ethereum/py-evm/blob/master/eth/db/account.py#L255). I'm hesitant about truncating because it could hide other potential errors.

I think ideally we have to tackle this at multiple levels, to prevent similar errors from happening:

- Make sure opcode handlers don't have to deal with invalid values by validating values in environment interface functions like `getContractStorage` (which SLOAD calls).
- Validate value sizes in the [stack](https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/vm/stack.js#L21), as `SSTORE` is getting a value larger than 32 bytes from the stack which I think shouldn't happen.
- Validate value size in data types like `Account`.

Many of these arise from the fact that we're using `BN`, which is arbitrary precision. I'll also try to look into how we can have fixed-width U256 numbers.

However these were all just my two cents, I'd appreciate feedback from others.

**Comment by holgerd77** [comment]
Think this sounds like a solid and sustainably tackling suggestion, @s1na did you lately mention https://github.com/ewasm/fixed-bn.js/? Not sure about the context I stumbled upon this...

**Comment by davidmurdoch** [comment]
@holgerd77, Going a bit off topic: but has there been any research into porting over to using native `BigInt` with [it's polyfill](https://github.com/GoogleChromeLabs/jsbi) in the vm and related libs? I've been considering it with Ganache but haven't done much research into it yet.

**Comment by s1na** [comment]
@holgerd77 yeah that was me. Created #470 to discuss this further.

**Comment by davidmurdoch** [comment]
@s1na 	I did some digging internally with the Truffle team and came up with the following trace of what happens (copy + pasted from out chat):

The issue is that when you send `value` to an address this happens:
https://github.com/ethereumjs/ethereumjs-vm/blob/c842d101108d6acdd9ca82ab1de274ba3b16e5b2/lib/runCall.js#L160-L170
which doesn't limit the size of resulting `balance`.

So when this happens later:
https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/vm/opFns.js#L306
...the value pushed on the stack can be > 32 bytes.

Which will lead to `SSTORE` actually setting that value in storage:
https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/vm/opFns.js#L540

Which may cause `MSTORE` to come around at some other point in time and do this:
https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/vm/opFns.js#L497

which will crash the vm and ganache when the value won't fit in the 32 byte Buffer it's allotted.

---

I hope this helps!

**Comment by s1na** [comment]
This is invaluable, thanks @davidmurdoch!

**Comment by holgerd77** [comment]
Do not have the complete overview, do we now already have enough minimal coverage of this with #471?

**Comment by holgerd77** [comment]
Without re-reading to grasp the whole context again, I would assume this has been solved? //cc @ryanio @jochem-brouwer

**Comment by jochem-brouwer** [comment]
Yes, closed by #880
